core/primitives-core/src/hash.rs

```diff
@@ -47,6 +46,31 @@ impl CryptoHash {
         let value = std::str::from_utf8(&buffer[..len]).unwrap();
         visitor(value)
     }
+
+    /// Decodes base58-encoded string into a 32-byte hash.
+    ///
+    /// Returns one of three results: success with the decoded CryptoHash,
+    /// invalid length error indicating that the encoded value was too short or
+    /// too long or other decoding error (e.g. invalid character).
+    fn from_base58_impl(encoded: &str) -> Decode58Result {
+        let mut result = Self::new();
+        match bs58::decode(encoded).into(&mut result.0) {
+            Ok(len) if len == result.0.len() => Decode58Result::Ok(result),
+            Ok(_) | Err(bs58::decode::Error::BufferTooSmall) => Decode58Result::BadLength,
+            Err(err) => Decode58Result::Err(err),
+        }
+    }
+}
+
+/// Result of decoding base58-encoded crypto hash.
+enum Decode58Result {
+    /// Decoding succeeded.
+    Ok(CryptoHash),
+    /// The decoded data has incorrect length; either too short or too long.
+    BadLength,
```

Comment:
`BadLength` is conceptually an error. Why does it need to be represented by a separate enum variant?

Reply:
Yes, if this was a public API I would have wrapped it into another enum or change things.  But for an internal API I found this to be most straightforward approach.